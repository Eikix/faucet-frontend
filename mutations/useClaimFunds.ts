import { useMutation } from "@tanstack/react-query";
import { API } from "@/lib/api";

interface ClaimFundsParams {
  walletAddress: string;
}

const useClaimFunds = () => {
  return useMutation({
    mutationKey: ["claimFunds"],
    mutationFn: (params: ClaimFundsParams) => API.faucet.claimFunds(params.walletAddress),
  });
};

export { useClaimFunds };
