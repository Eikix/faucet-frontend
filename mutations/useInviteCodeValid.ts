import { useMutation } from "@tanstack/react-query";
import { API } from "@/lib/api";

interface CheckInviteCodeParams {
  inviteCode: string;
}

// NOTE: mutations are generally used for POST, PUT, DELETE, etc., --  you can technically use them for GET requests if you need to trigger them manually or handle them differently from typical data fetching scenarios. However, itâ€™s crucial to consider the semantic meaning of mutations in the context of APIs and maintain consistency with how these operations are typically handled.
const useInviteCodeValid = () => {
  return useMutation({
    mutationKey: ["inviteCodeValid"],
    mutationFn: (params: CheckInviteCodeParams) => API.invite.isValid(params.inviteCode),
  });
};

export { useInviteCodeValid };
