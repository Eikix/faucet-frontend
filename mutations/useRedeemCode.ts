import { useMutation } from "@tanstack/react-query";
import { API } from "@/lib/api";

interface RedeemCodeParams {
  inviteCode: string;
  walletAddress: string;
}

const useRedeemCode = () => {
  return useMutation({
    mutationKey: ["redeemCode"],
    mutationFn: (params: RedeemCodeParams) => API.invite.redeemCode(params.inviteCode, params.walletAddress),
  });
};

export default useRedeemCode;
