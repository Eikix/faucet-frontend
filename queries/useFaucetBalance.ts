import { useQuery } from "@tanstack/react-query";
import { API } from "@/lib/api";

const useFaucetBalance = () => {
  return useQuery({ queryKey: ["faucetBalance"], queryFn: () => API.faucet.getBalance(), refetchInterval: false });
};

export { useFaucetBalance };
