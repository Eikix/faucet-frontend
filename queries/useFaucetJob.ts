import { useRef } from "react";
import { useQuery } from "@tanstack/react-query";
import { API } from "@/lib/api";

interface FaucetJobParams {
  jobId: string;
}

const isComplete = (query: any, timePassed: number) =>
  (query.state.data && query.state.data[0].status === "completed") || timePassed > 60;

const useFaucetJob = (jobId: string) => {
  const countRef = useRef(0);
  return useQuery({
    queryKey: ["useFaucetJob", jobId],
    queryFn: () => {
      countRef.current += 1;
      return API.jobs.claimFunds(jobId);
    },
    refetchInterval: (query) => (isComplete(query, countRef.current) ? false : 1000),
  });
};

export { useFaucetJob };
