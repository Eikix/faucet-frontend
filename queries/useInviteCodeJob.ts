import { useQuery } from "@tanstack/react-query";
import { API } from "@/lib/api";

interface InviteCodeJobParams {
  jobId: string;
}

const isComplete = (query: any) => query.state.data && query.state.data[0].status === "completed";

const useInviteCodeJob = (jobId: string) => {
  return useQuery({
    queryKey: ["useInviteCodeJob", jobId],
    queryFn: () => API.jobs.redeemCode(jobId),
    refetchInterval: (query) => (isComplete(query) ? false : 2000),
  });
};

export { useInviteCodeJob };
