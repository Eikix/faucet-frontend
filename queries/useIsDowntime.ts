import { useQuery } from "@tanstack/react-query";
import { API } from "@/lib/api";

const useIsDowntime = () => {
  return useQuery({
    queryKey: ["useIsDowntime"],
    queryFn: () => API.general.isDowntime(),
    refetchInterval: 1000 * 10, // 10 seconds
  });
};

export { useIsDowntime };
