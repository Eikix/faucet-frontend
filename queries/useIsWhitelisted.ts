import { useQuery } from "@tanstack/react-query";
import { API } from "@/lib/api";

interface IsWhitelistedParams {
  address: string;
}

const useIsWhitelisted = (address: string) => {
  return useQuery({
    queryKey: ["useIsWhitelisted", address],
    queryFn: () => API.general.isWhitelisted(address),
  });
};

export { useIsWhitelisted };
