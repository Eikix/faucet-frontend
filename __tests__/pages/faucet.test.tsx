import "@testing-library/jest-dom";
import { act, render, screen, waitFor } from "@testing-library/react";
import { useRouter } from "next/navigation";
import { useFaucet } from "@/hooks/useFaucet";
import { useBlockNumber, useWalletBalance, lightTheme } from "thirdweb/react";
import { useWindowSize } from "@/hooks/useWindowSize";
import { useFaucetJob } from "@/queries/useFaucetJob";
import { useClaimFunds } from "@/mutations/useClaimFunds";
import { toast } from "sonner";
import Faucet from "@/app/faucet/page";

// Mock hooks and components
jest.mock("next/navigation", () => ({
  useRouter: jest.fn(),
}));
jest.mock("@/hooks/useFaucet", () => ({
  useFaucet: jest.fn(),
}));
jest.mock("thirdweb/react", () => ({
  useBlockNumber: jest.fn(),
  useWalletBalance: jest.fn(),
  lightTheme: jest.fn(() => ({ colors: {} })), // Mock lightTheme function
}));
jest.mock("@/hooks/useWindowSize", () => ({
  useWindowSize: jest.fn(),
}));
jest.mock("@/queries/useFaucetJob", () => ({
  useFaucetJob: jest.fn(),
}));
jest.mock("@/mutations/useClaimFunds", () => ({
  useClaimFunds: jest.fn(),
}));
jest.mock("sonner", () => ({
  toast: {
    message: jest.fn(),
    error: jest.fn(),
  },
}));

// Mock components
jest.mock("react-confetti", () => () => <div />);
jest.mock("@/components/faucet-claim", () => ({
  FaucetClaim: ({
    handleClaim,
    isProcessing,
    isCooldown,
    isOutOfFunds,
    available,
    faucetStats,
  }: {
    handleClaim: () => void;
    isProcessing: boolean;
    isCooldown: boolean;
    isOutOfFunds: boolean;
    available: string;
    faucetStats: {};
  }) => (
    <div>
      <button
        onClick={handleClaim}
        disabled={isProcessing}
        className={!isProcessing && (isCooldown || isOutOfFunds) ? "hidden" : ""}
      >
        {isProcessing ? "Claiming.." : isCooldown ? "Cooldown" : "Claim"}
      </button>
    </div>
  ),
}));
jest.mock("@/components/faucet-success", () => ({
  FaucetSuccess: ({ navigateToClaim, txHash }: { navigateToClaim: () => void; txHash: string }) => (
    <div>
      <button onClick={navigateToClaim}>Back to Claim</button>
      <a href={`https://explorer.kakarot.io/tx/${txHash}`} rel="noopener noreferrer" target="_blank">
        View on Explorer
      </a>
    </div>
  ),
}));

jest.mock("@/components/ui/skeleton", () => ({
  Skeleton: () => <div>Skeleton Loading...</div>,
}));

describe("Faucet Page", () => {
  const mockReplace = jest.fn();
  const mockUseRouter = useRouter as jest.Mock;
  const mockUseFaucet = useFaucet as jest.Mock;
  const mockUseBlockNumber = useBlockNumber as jest.Mock;
  const mockUseWalletBalance = useWalletBalance as jest.Mock;
  const mockUseWindowSize = useWindowSize as jest.Mock;
  const mockUseFaucetJob = useFaucetJob as jest.Mock;
  const mockUseClaimFunds = useClaimFunds as jest.Mock;

  beforeEach(() => {
    mockUseRouter.mockReturnValue({ replace: mockReplace });
    mockUseFaucet.mockReturnValue({
      wallet: null,
      faucetStats: null,
      faucetBalance: null,
      refetchFaucet: jest.fn(),
      isFaucetLoading: false,
      isWhitelisted: false,
    });
    mockUseBlockNumber.mockReturnValue(null);
    mockUseWalletBalance.mockReturnValue({ refetch: jest.fn() });
    mockUseWindowSize.mockReturnValue({ width: 1024 });
    mockUseFaucetJob.mockReturnValue({ data: null, isError: false });
    mockUseClaimFunds.mockReturnValue({ mutate: jest.fn(), isPending: false, data: null });
    jest.clearAllMocks();
  });

  describe("Loading State", () => {
    it("renders SkeletonLoader when isFaucetLoading is true", () => {
      mockUseFaucet.mockReturnValue({ ...mockUseFaucet(), isFaucetLoading: true });

      render(<Faucet />);

      expect(screen.getAllByText("Skeleton Loading...").length).toBeGreaterThan(0);
    });
  });

  describe("Redirects", () => {
    it("redirects to home page when wallet is not available", () => {
      mockUseFaucet.mockReturnValue({ ...mockUseFaucet(), wallet: null });

      render(<Faucet />);

      expect(mockReplace).toHaveBeenCalledWith("/");
    });

    it("redirects to invite-code page when not whitelisted", () => {
      mockUseFaucet.mockReturnValue({ ...mockUseFaucet(), isWhitelisted: false });

      render(<Faucet />);

      expect(mockReplace).toHaveBeenCalledWith("/invite-code");
    });
  });

  describe("Normal Render", () => {
    beforeEach(() => {
      mockUseFaucet.mockReturnValue({
        wallet: { address: "0x123" },
        faucetStats: { dripAmountInEth: "0.001", timeLeftInS: 0, canClaim: true },
        faucetBalance: { faucetBalanceInEth: "1" },
        refetchFaucet: jest.fn(),
        isFaucetLoading: false,
        isWhitelisted: true,
      });
      mockUseBlockNumber.mockReturnValue("12345");
    });

    it("renders main content when wallet is available and whitelisted", () => {
      render(<Faucet />);

      expect(screen.getByRole("heading", { level: 3 })).toHaveTextContent("Need more testnet ETH?");
      expect(
        screen.getByText("Reach out to us on Discord and raise a ticket if you need large amount of testnet ETH.")
      ).toBeInTheDocument();
      expect(screen.getByText("Claim")).toBeInTheDocument();
    });

    it("shows faucet balance and block number", () => {
      render(<Faucet />);

      expect(screen.getByText("Faucet Balance:")).toBeInTheDocument();
      expect(screen.getByText("1ETH")).toBeInTheDocument();
      expect(screen.getByText("Block Number:")).toBeInTheDocument();
      expect(screen.getByText("12345")).toBeInTheDocument();
    });

    it("handles claim process and shows success state", async () => {
      const refetchFaucet = jest.fn();
      const refetchWallet = jest.fn();
      mockUseFaucet.mockReturnValue({
        wallet: { address: "0x123" },
        faucetStats: { dripAmountInEth: "0.001", timeLeftInS: 0, canClaim: true },
        faucetBalance: { faucetBalanceInEth: "1" },
        refetchFaucet,
        isFaucetLoading: false,
        isWhitelisted: true,
      });
      mockUseClaimFunds.mockReturnValue({
        mutate: jest.fn((params, { onSuccess }) => onSuccess({ jobID: "job123" })),
        isPending: false,
        data: null,
      });
      mockUseFaucetJob.mockReturnValue({ data: [{ status: "completed", transaction_hash: "0xabc" }], isError: false });
      mockUseWalletBalance.mockReturnValue({ refetch: refetchWallet });

      render(<Faucet />);

      await waitFor(() => {
        expect(screen.getByText("Back to Claim")).toBeInTheDocument();
      });

      expect(refetchFaucet).toHaveBeenCalled();
      expect(refetchWallet).toHaveBeenCalled();
      expect(toast.message).toHaveBeenCalledWith(expect.anything());
    });

    it("handles error state during claim process", async () => {
      mockUseFaucetJob.mockReturnValue({ data: null, isError: true });

      render(<Faucet />);

      await waitFor(() => {
        expect(toast.error).toHaveBeenCalledWith("Failed to claim funds. Please try again later.");
      });
    });
  });

  describe("Edge Cases", () => {
    it("handles case when faucetStats are not available", () => {
      mockUseFaucet.mockReturnValue({ ...mockUseFaucet(), faucetStats: null });

      render(<Faucet />);

      expect(screen.getByText("Claim")).toBeInTheDocument();
    });

    it("handles case when faucet balance is low", () => {
      mockUseFaucet.mockReturnValue({ ...mockUseFaucet(), faucetBalance: { faucetBalanceInEth: "0.000005" } });

      render(<Faucet />);

      expect(screen.getByText("Claim")).toBeInTheDocument();
    });

    it("handles cooldown state", () => {
      mockUseFaucet.mockReturnValue({
        ...mockUseFaucet(),
        faucetStats: { dripAmountInEth: "0.001", timeLeftInS: 60, canClaim: false },
      });

      render(<Faucet />);

      expect(screen.getByText("Cooldown")).toBeInTheDocument();
    });

    it("updates correctly when useFaucet value changes", () => {
      mockUseFaucet.mockReturnValue({ isFaucetLoading: true, isWhitelisted: false, wallet: null });

      const { rerender } = render(<Faucet />);
      expect(screen.getAllByText("Skeleton Loading...").length).toBeGreaterThan(0);

      act(() => {
        mockUseFaucet.mockReturnValue({
          isFaucetLoading: false,
          isWhitelisted: false,
          wallet: { address: "0x123" },
          faucetStats: { dripAmountInEth: "0.001", timeLeftInS: 0, canClaim: true },
          faucetBalance: { faucetBalanceInEth: "1" },
        });
      });

      rerender(<Faucet />);
      expect(screen.getByText("Claim")).toBeInTheDocument();
    });
  });

  describe("Snapshot Test", () => {
    it("matches snapshot", () => {
      mockUseFaucet.mockReturnValue({
        isFaucetLoading: false,
        isWhitelisted: true,
        wallet: { address: "0x123" },
        faucetStats: { dripAmountInEth: "0.001", timeLeftInS: 0, canClaim: true },
        faucetBalance: { faucetBalanceInEth: "1" },
      });

      const { asFragment } = render(<Faucet />);

      expect(asFragment()).toMatchSnapshot();
    });
  });
});
