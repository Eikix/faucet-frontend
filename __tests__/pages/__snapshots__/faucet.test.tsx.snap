// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Faucet Page Snapshot Test matches snapshot 1`] = `
<DocumentFragment>
  <main
    class="flex flex-col items-center mt-10 h-full"
  >
    <div
      class="flex flex-col bg-white w-full py-6 px-3 sm:px-10 lg:px-20 rounded-md mb-12"
    >
      <div />
      <div
        class="flex flex-col sm:flex-row gap-4 sm:gap-0 justify-between"
      >
        <h4
          class="text-[#878794] faucetDetails px-3 py-2 rounded-sm"
        >
          Faucet Balance: 
          <span
            class="text-[#f54400]"
          >
            1ETH
          </span>
        </h4>
        <h4
          class="text-[#878794] faucetDetails px-3 py-2 rounded-sm"
        >
          Block Number: 
          <span
            class="text-[#f54400]"
          >
            0x
          </span>
        </h4>
      </div>
      <div>
        <button
          class=""
        >
          Claim
        </button>
      </div>
    </div>
    <section
      class="flex flex-col justify-center items-center px-4"
    >
      <header>
        <h3
          class="scroll-m-20 text-2xl tracking-tight"
        >
          Need more testnet ETH?
        </h3>
      </header>
      <p
        class="leading-5 [&:not(:first-child)]:mt-4 text-[#878794] text-center text-pretty max-w-[400px]"
      >
        Reach out to us on Discord and raise a ticket if you need large amount of testnet ETH.
      </p>
    </section>
    <a
      class="pb-10"
      href="https://discord.gg/kakarotzkevm"
      rel="noopener noreferrer"
      target="_blank"
    >
      <button
        class="inline-flex items-center justify-center whitespace-nowrap text-sm transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground font-normal p-3 rounded-md w-max space-x-2 max-w-[120px] mt-6"
      >
        <span>
          Reach Out
        </span>
        <img
          alt="Docs"
          data-nimg="1"
          decoding="async"
          height="16"
          loading="lazy"
          src="/assets/link-icon.svg"
          style="color: transparent;"
          width="16"
        />
      </button>
    </a>
  </main>
</DocumentFragment>
`;
